### DISCUSSION TITLE
Concepts for Efficient Coding (Analysis of Algorithms)

### DISCUSSION THREAD

**Discussion by liamjberrisford** [coding-for-reproducible-research/CfRR_Courses#128] [Workshop Suggestions]
# Concepts for Efficient Coding

## Course Description
This course aims to provide an intuitive and deeper understanding of what happens when you run code. By focusing on the underlying computer architecture and the principles of data structures and algorithms, participants will gain the knowledge needed to make informed choices in their coding practices. The course will not teach specific implementations but will equip students with the confidence to understand and learn various tools and techniques to optimize code performance.

## Course Objectives
- Understand the fundamentals of computer architecture.
- Learn the importance of data structures and algorithms in coding efficiency.
- Gain an understanding of the basic process for data science from data storage to computation.
- Explore the implications of code on different levels of memory hierarchy (CPU Cache, RAM, Disk).
- Develop the ability to analyse algorithms focusing on running time and memory space.
- Understand the differences between strongly and weakly typed languages, and compiled and interpreted code.
- Learn to reduce search space and apply inventive solutions for efficient coding.
- Understand the concept of vectorization and parallel computation.
- Gain practical insights into building computation graphs and optimizing code performance.

## Prerequisites
- Basic knowledge of programming concepts.
- Familiarity with Python or some other programming language

**Comment by ejh243** [comment]
Would this build on the content from the computational thinking course? Or be an intermediate version of that?

**Comment by liamjberrisford** [comment]
I was thinking of building it upon the computational thinking course but for it to be slightly more mathematical and technical in focus. I've attached a PDF of the slides I used when delivering a similar workshop to PhD students. 

https://github.com/berrli/ConceptsForEfficientProgramming/blob/main/Aux%20CDT%20Presentation%20-%20Concepts%20for%20Efficient%20Coding.pptx
